File: benchmarks3d/many_kinematics3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Create the balls
     */
    let num = 30;
    let rad = 1.0;

    let shift = rad * 6.0 + 1.0;
    let centerx = shift * (num as f32) / 2.0;
    let centery = shift * (num as f32) / 2.0;
    let centerz = shift * (num as f32) / 2.0;

    for i in 0..num {
        for j in 0usize..num {
            for k in 0..num {
                let x = i as f32 * shift - centerx;
                let y = j as f32 * shift - centery;
                let z = k as f32 * shift - centerz;

                // Build the rigid body.
                let velocity = Vector::new(
                    rand::random::<f32>() - 0.5,
                    rand::random::<f32>() - 0.5,
                    rand::random::<f32>() - 0.5,
                ) * 30.0;
                let rigid_body = RigidBodyBuilder::new(RigidBodyType::KinematicVelocityBased)
                    .translation(vector![x, y, z])
                    .linvel(velocity);
                let handle = bodies.insert(rigid_body);
                let collider = ColliderBuilder::ball(rad);
                colliders.insert_with_parent(collider, handle, &mut bodies);
            }
        }
    }

    testbed.add_callback(move |_, physics, _, _| {
        for (_, rb) in physics.bodies.iter_mut() {
            let mut linvel = *rb.linvel();

            for dim in 0..3 {
                if (linvel[dim] > 0.0 && rb.translation()[dim] > (shift * num as f32) / 2.0)
                    || (linvel[dim] < 0.0 && rb.translation()[dim] < -(shift * num as f32) / 2.0)
                {
                    linvel[dim] = -linvel[dim];
                }
            }

            rb.set_linvel(linvel, false);
        }
    });

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![100.0, 100.0, 100.0], Point::origin());
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/stress_tests/many_kinematics3.py
"""Port of examples3d/stress_tests/many_kinematics3.rs."""
from __future__ import annotations

import random

import rapier3d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "Many kinematics"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    num = 30
    rad = 1.0
    shift = rad * 6.0 + 1.0
    centerx = shift * num / 2.0
    centery = shift * num / 2.0
    centerz = shift * num / 2.0

    rng = random.Random(0)
    half = shift * num / 2.0
    for i in range(num):
        for j in range(num):
            for k in range(num):
                x = i * shift - centerx
                y = j * shift - centery
                z = k * shift - centerz
                velocity = (
                    (rng.random() - 0.5) * 30.0,
                    (rng.random() - 0.5) * 30.0,
                    (rng.random() - 0.5) * 30.0,
                )
                body = (
                    rp.RigidBody.kinematic_velocity_based()
                    .translation((x, y, z))
                    .linvel(velocity)
                )
                h = bodies.insert(body)
                colliders.insert_with_parent(rp.Collider.ball(rad), h, bodies)

    def _cb(tb) -> None:
        for _, rb in tb.bodies:
            lv = rb.linvel
            new_x = lv.x
            new_y = lv.y
            new_z = lv.z
            t = rb.translation
            if (new_x > 0.0 and t.x > half) or (new_x < 0.0 and t.x < -half):
                new_x = -new_x
            if (new_y > 0.0 and t.y > half) or (new_y < 0.0 and t.y < -half):
                new_y = -new_y
            if (new_z > 0.0 and t.z > half) or (new_z < 0.0 and t.z < -half):
                new_z = -new_z
            rb.linvel = rp.Vec3(new_x, new_y, new_z)

    testbed.add_callback(_cb)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
